src/tactic/interactive_expr.lean

```diff
@@ -65,6 +67,80 @@ meta def sf.of_eformat : eformat → sf
 | (of_format f) := sf.of_string $ format.to_string f
 | (compose x y) := sf.compose (sf.of_eformat x) (sf.of_eformat y)
 
+/-- `subtract x y` is some `z` when `∃ z, x = y ++ z` [fixme] this probably already exists. -/
+def subtract {α} [decidable_eq α]: (list α) → (list α) → option (list α)
+| a [] := some a -- [] ++ a = a
+| [] _ := none   -- (h::t) ++ _ ≠ []
+-- if t₂ ++ z = t₁ then (h₁ :: t₁) ++ z = (h₁ :: t₂)
+| (h₁ :: t₁) (h₂ :: t₂) := if h₁ = h₂ then subtract t₁ t₂ else none
+
+/-- Get the subexpression and sf object at the given address, if any.
+The address must point to a `sf.tag_expr` boundary. -/
+meta def sf.follow : expr.address → sf → option (expr × sf)
+| [] s := none
+| l (sf.tag_expr ea e  m) := do
+  a ← subtract l ea,
+  if a = [] then pure (e,m) else
+  sf.follow a m
+| l (sf.block _ a) := sf.follow l a
+| l (sf.highlight _ a) := sf.follow l a
+| l (sf.of_string _) := none
+| l (sf.compose a b) := sf.follow l a <|> sf.follow l b
+
+/-- Run `f` on each of the immediate child substrings of the given string. -/
+meta def sf.traverse {m} [applicative m] (f : sf → m sf) : sf → m sf
+| (sf.tag_expr ea e m) := pure (sf.tag_expr ea e) <*> f m
+| (sf.block x a) := pure (sf.block x) <*> f a
+| (sf.highlight x a) := pure (sf.highlight x) <*> f a
+| (sf.compose a b) := pure (sf.compose) <*> f a <*> f b
+| (sf.of_string s) := pure $ sf.of_string s
+
+/-- Run `f` on each child. If it fails then stop traversing and keep it the same.-/
+meta def sf.replace {m} [monad m] [alternative m] (f : sf → m sf) : sf → m sf
+| x := (f x >>= sf.traverse sf.replace) <|> pure x
+
+/-- The test for whether the proposition is a valid target for restricted quantifier collapsing.
```

Comment:
Aren't the valid targets "everything with infix notation"?

Reply:
Is there a programmatic way to test whether a given relation has an infix notation?

Reply:
Not that I'm aware of. You could build a dummy application and see if it pretty prints differently than you expect, which is very hackish.